### ISSUE TITLE
What is Model-Agnostic Meta-learning (MAML) ?

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#345]
https://towardsdatascience.com/model-agnostic-meta-learning-maml-8a245d9bc4ac

**Comment by sweaterr** [comment]
Artificial intelligence is trying to learn how to learn from the way humans learn.
인공 지능은 인간이 배우는 방식에서 배우는 법을 배우려고합니다.

We can quickly and easily recognize a new object from just seeing one or few pictures of it, or even from only reading about it without having ever seen it before.
우리는 새로운 물체를 단지 하나 또는 몇 장의 사진을 보거나 심지어 이전에 본 적이없는 독서만으로도 빠르고 쉽게 인식 할 수 있습니다.

We can learn quickly a new skill as well as master many different tasks.
우리는 새로운 기술을 빨리 배울 수있을뿐만 아니라 여러 가지 다른 일을 마스터 할 수 있습니다.

This seems easy for the human intelligence but for machines, it is quite a challenge to overcome.
이것은 인간의 지능으로는 쉽게 보이지만 기계의 경우에는 극복하기가 어렵습니다.

Berkeley AI Research Lab published a research paper introducing Model-Agnostic Meta-learning MAML which is a simple solution, yet so powerful and game-changing in meta-learning (or learning to learn).
버클리 인공 지능 연구실 (BAM)은 단순한 솔루션 인 메타 신디케이션 메타 러닝 MAML을 소개하는 연구 보고서를 발표했으나 메타 러닝 (또는 학습을 배우는)에서 강력하고 게임을 변화 시켰습니다.

The objective of this article is to give you a clear understanding of this approach.
이 기사의 목적은이 접근 방식을 명확하게 이해하는 것입니다.

**Comment by sweaterr** [comment]
## Few-Shot Meta-learning

Deep learning has a great success in mastering one task using a large dataset.
딥 학습은 대규모 데이터 세트를 사용하여 하나의 작업을 마스터하는 데 큰 성공을 거둔다.

But, what we really want to achieve is few shot-meta learning which is an algorithm that trains a neural network to learn many different tasks using only a small data per task.
그러나 우리가 실제로 달성하기를 원하는 것은 숏 - 메타 학습 (task-meta learning)입니다.이 알고리즘은 신경망을 훈련시켜 작업마다 작은 데이터만을 사용하여 여러 가지 다양한 작업을 학습하는 알고리즘입니다.

In meta-learning, there is a meta-learner and a learner.
메타 학습에는 메타 학습자와 학습자가 있습니다.

The meta-learner (or the agent) trains the learner (or the model) on a training set that contains a large number of different tasks.
메타 학습자 (또는 에이전트)는 학습자 (또는 모델)에게 많은 수의 다양한 작업이 포함 된 교육 세트를 교육합니다.

In this stage of meta-learning, the model will acquire a prior experience from training and will learn the common features representations of all the tasks.
메타 학습의이 단계에서 모델은 사전 교육 경험을 습득하고 모든 작업의 ​​공통된 특징을 학습합니다.

Then, whenever, there is a new task to learn, the model with its prior experience will be fine-tuned using the small amount of the new training data brought by that task.
그런 다음 학습 할 새로운 작업이있을 때마다 이전 경험을 가진 모델은 해당 작업으로 가져온 소량의 새로운 교육 데이터를 사용하여 미세 조정됩니다.

But we don’t want to start from a random initialization of its parameters because if we do so, it will not converge to a good performance after only a few updates on each task.
그러나 매개 변수의 무작위 초기화로 시작하고 싶지는 않습니다. 왜냐하면 그렇게 할 경우 각 작업에 대한 업데이트 만 수행하면 좋은 성능으로 수렴되지 않기 때문입니다.

**Comment by sweaterr** [comment]
## So how can we get an optimal starting point for fine-tuning?

Model-Agnostic Meta-Learning (MAML) provides a good initialization of a model’s parameters to achieve an optimal fast learning on a new task with only a small number of gradient steps while avoiding overfitting that may happen when using a small dataset.
Model-Agnostic Meta-Learning (MAML)은 작은 데이터 세트를 사용할 때 발생할 수있는 과잉을 피하면서 적은 수의 그래디언트 단계만으로 새로운 작업에 대한 최적의 빠른 학습을 달성하기 위해 모델 매개 변수의 좋은 초기화를 제공합니다.

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/38785867-e622a798-415e-11e8-951a-9b854bf4ef00.png)

**Comment by sweaterr** [comment]
In the diagram above, θ is the model’s parameters and the bold black line is the meta-learning phase.
위의 다이어그램에서 θ는 모델의 매개 변수이고 굵게 표시된 검은 선은 메타 학습 단계입니다.

When we have, for example, 3 different new tasks 1, 2 and 3, a gradient step is taken for each task (the gray lines).
예를 들어 3 가지 새로운 작업 1, 2 및 3이있는 경우 각 작업 (회색 선)에 대한 단계 그라디언트 단계가 수행됩니다.

We can see that the parameters θ are close to all the 3 optimal parameters of task 1, 2, and 3 which makes θ the best parameters initialization that can quickly adapt to different new tasks.
매개 변수 θ가 작업 1, 2 및 3의 3 가지 최적 매개 변수에 모두 가까워서 θ를 다른 새로운 작업에 빠르게 적응할 수있는 최상의 매개 변수 초기화로 만듭니다.

As a result, only a small change in the parameters θ will lead to an optimal minimization of the loss function of any task.
결과적으로 매개 변수 θ의 작은 변화만으로 모든 작업의 ​​손실 함수를 최적으로 최소화 할 수 있습니다.

**Comment by sweaterr** [comment]
As we can see, the meta-gradient update contains a gradient through a gradient which can be computed using the Hessian-vector product.
알 수 있듯이 메타 그래디언트 업데이트에는 Hessian 벡터 제품을 사용하여 계산할 수있는 그래디언트를 통한 그래디언트가 포함되어 있습니다.

Then, we repeat the same process until we train all the batches.
그런 다음, 우리는 모든 배치를 훈련 할 때까지 동일한 과정을 반복합니다.

**Comment by sweaterr** [comment]
## Compatibility

MAML can work on any model trained with a gradient descent including fully connected and convolutional networks, and it can be used with a variety of loss functions.
MAML은 완전 연결 및 컨벌루션 네트워크를 포함하여 그라디언트 디센트로 교육 된 모든 모델에서 작동 할 수 있으며 다양한 손실 기능과 함께 사용할 수 있습니다.

What is really important, is its compatibility with different domains such as few-shot regression, image classification, and reinforcement learning, where it was able to outperform a number of existing approaches without expanding the parameters number.
정말로 중요한 점은 소수점 회귀, 이미지 분류 및 강화 학습과 같은 다른 영역과의 호환성 때문입니다. 매개 변수 번호를 확장하지 않고도 기존의 여러 가지 방법보다 우수한 성능을 보였습니다.

**Comment by sweaterr** [comment]
## Meta-Learning and Universality

The most popular approach to meta-learning is using a Recurrent Neural Network RNN as a meta-learner to train another model.
메타 학습에 대한 가장 보편적 인 접근 방법은 다른 모델을 훈련시키는 메타 학습자로서 Recurrent Neural Network RNN을 사용하는 것입니다.

For instance, (Ravi & Larochelle, 2017) proposed an LSTM based meta-learner to train a classifier in few-shot learning to get the appropriate parameters and a general initialization of these parameters.
예를 들어 (Ravi & Larochelle, 2017)는 LSTM 기반의 메타 학습자 (meta-learner)를 사용하여 적절한 매개 변수를 얻고 이러한 매개 변수를 일반적으로 초기화하는 소수 학습에서 분류자를 교육했습니다.

However, according to (Finn & Levine, 2017 ) written by the same authors of the MAMl paper, MAML is able to approximate any learning algorithm compared to recurrent learners.The MAML initializations are more resilient to overfitting to small datasets, and they are also more effective when the model is dealing with new tasks that were not in the training set.
그러나 MAMl 논문의 동일한 저자가 작성한 (Finn & Levine, 2017)에 따르면 MAML은 반복 학습자와 비교하여 모든 학습 알고리즘을 근사 할 수 있습니다. MAML 초기화는 작은 데이터 집합에 과도하게 탄력적이며 모델이 교육 세트에 포함되지 않은 새 작업을 처리 할 때보다 효과적입니다.

**Comment by sweaterr** [comment]
## References
Chelsea Finn, Pieter Abbeel, and Sergey Levine. Model-agnostic meta-learning for fast adaptation of deep networks. arXiv preprint arXiv:1703.03400, 2017.

Ravi, Sachin and Larochelle, Hugo. Optimization as a model for few-shot learning. (ICLR), 2017.

Chelsea Finn and Sergey Levine. Meta-learning and universality: Deep representations and gradient descent can approximate any learning algorithm. arXiv preprint arXiv:1710.11622, 2017.
